### PR TITLE
ICE fixes and other things

### PR DESCRIPTION
**Pull request by Manishearth**
h/t @eddyb for help with the region stuff

### CODE REVIEW

**Comment by Manishearth** [comment]
Merging now since this is breaking people, feel free to post-review

File: src/utils/mod.rs [hunk 1]
```diff
@@ -764,8 +776,13 @@ pub fn unsugar_range(expr: &Expr) -> Option<UnsugaredRange> {
 }
 
 /// Convenience function to get the return type of a function or `None` if the function diverges.
-pub fn return_ty(fun: ty::Ty) -> Option<ty::Ty> {
-    if let ty::FnConverging(ret_ty) = fun.fn_sig().skip_binder().output {
+pub fn return_ty<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, fn_item: NodeId) -> Option<ty::Ty<'tcx>> {
+    let parameter_env = ty::ParameterEnvironment::for_item(cx.tcx, fn_item);
+    let infcx = infer::new_infer_ctxt(cx.tcx, &cx.tcx.tables, Some(parameter_env), ProjectionMode::Any);
```
Comment (eddyb) [inline src/utils/mod.rs]: I don't think you need the infcx here, just in `same_tys`.

Comment (Manishearth) [inline src/utils/mod.rs]: we were unwrapping the binder here

Comment (eddyb) [inline src/utils/mod.rs]: I mean, you can use the `ParameterEnvironment`, and just not create an `infcx` out of it.

Comment (Manishearth) [inline src/utils/mod.rs]: good point, fixing

**Pull request merged**